File: app/src/test/tests/buttons.ts
/* buttons — function-exists audit + light behavior contracts.
 * Legacy reference: runButtonAudit() (lines 12958-12973) + testButton_*
 * / testTen_* / testCrit_* (~lines 13380-14250). Behaviour tests save
 * state before calling, assert the expected mutation, then restore
 * (legacy "save/restore" pattern).
 */
import { BUTTON_REGISTRY } from '../registry';
import type { TestResult, TestCheck } from '../types';
import {
  cart,
  categoryPath,
  menuOpen,
  searchOpen,
  toggleMenu,
  toggleSearch,
  drillInto,
  resetCategory,
  setQty,
  qtyOf,
  goUp,
} from '../../store/app-store';
import {
  bsOpen,
  toggleBs,
  activePersona,
  setPersona,
} from '../../store/bs-store';
import { childrenOf } from '../../data/catalog';

function existenceCheck(): TestResult[] {
  return BUTTON_REGISTRY.map((entry) => ({
    id: `button:${entry.fn}`,
    category: 'buttons' as const,
    label: entry.does,
    area: entry.area,
    checks: [
      {
        name: `${entry.fn} מוגדר וניתן להזעיק`,
        pass: typeof entry.ref === 'function',
        expected: 'function',
        got: typeof entry.ref,
      },
    ],
  }));
}

function behaviorCheck(): TestResult[] {
  const results: TestResult[] = [];

  /* toggleBs — flips bsOpen, twice restores */
  results.push(behaviorOne('toggleBs', 'מצב BS dial מתחלף', () => {
    const checks: TestCheck[] = [];
    const before = bsOpen.value;
    toggleBs();
    checks.push(verdict('מצב התהפך', bsOpen.value !== before, !before, bsOpen.value));
    toggleBs();
    checks.push(verdict('שתי קריאות חוזרות למצב המקורי', bsOpen.value === before, before, bsOpen.value));
    return checks;
  }));

  /* toggleMenu */
  results.push(behaviorOne('toggleMenu', 'מצב התפריט מתחלף', () => {
    const checks: TestCheck[] = [];
    const before = menuOpen.value;
    toggleMenu();
    checks.push(verdict('מצב התהפך', menuOpen.value !== before, !before, menuOpen.value));
    toggleMenu();
    checks.push(verdict('שתי קריאות חוזרות למצב המקורי', menuOpen.value === before, before, menuOpen.value));
    return checks;
  }));

  /* toggleSearch */
  results.push(behaviorOne('toggleSearch', 'מצב החיפוש מתחלף', () => {
    const checks: TestCheck[] = [];
    const before = searchOpen.value;
    toggleSearch();
    checks.push(verdict('מצב התהפך', searchOpen.value !== before, !before, searchOpen.value));
    toggleSearch();
    checks.push(verdict('שתי קריאות חוזרות למצב המקורי', searchOpen.value === before, before, searchOpen.value));
    return checks;
  }));

  /* setPersona — switch and restore */
  results.push(behaviorOne('setPersona', 'החלפת persona מעדכנת state', () => {
    const checks: TestCheck[] = [];
    const before = activePersona.value;
    const target = before === 'manager' ? 'contractor' : 'manager';
    setPersona(target);
    checks.push(verdict('activePersona מתעדכן', activePersona.value === target, target, activePersona.value));
    setPersona(before);
    checks.push(verdict('שחזור למצב המקורי', activePersona.value === before, before, activePersona.value));
    return checks;
  }));

  /* drillInto + resetCategory + goUp — navigate, then restore */
  results.push(behaviorOne('drillInto', 'כניסה לקטגוריה משרשרת את הנתיב', () => {
    const checks: TestCheck[] = [];
    const beforePath = [...categoryPath.value];
    const tops = childrenOf(null);
    if (tops.length === 0) {
      checks.push({ name: 'יש קטגוריות שורש', pass: false, detail: 'CATEGORIES.length === 0' });
      return checks;
    }
    resetCategory();
    drillInto(tops[0]!.id);
    checks.push(verdict('הנתיב גדל ב-1', categoryPath.value.length === 1, '1', String(categoryPath.value.length)));
    checks.push(verdict('הנתיב מכיל את הקטגוריה הנבחרת', categoryPath.value[0] === tops[0]!.id, tops[0]!.id, categoryPath.value[0] ?? ''));
    goUp();
    checks.push(verdict('goUp מקצר את הנתיב', categoryPath.value.length === 0, '0', String(categoryPath.value.length)));
    /* restore */
    for (const id of beforePath) drillInto(id);
    return checks;
  }));

  /* setQty / qtyOf — set, read, restore */
  results.push(behaviorOne('setQty', 'הגדרת כמות בעגלה משתקפת ב-qtyOf', () => {
    const checks: TestCheck[] = [];
    const beforeCart = cart.value.map((l) => ({ ...l }));
    const testId = '__regression_test_product__';
    setQty(testId, 3);
    checks.push(verdict('qtyOf מחזיר את הערך החדש', qtyOf(testId) === 3, '3', String(qtyOf(testId))));
    setQty(testId, 0);
    checks.push(verdict('הגדרת 0 מסירה את הפריט', qtyOf(testId) === 0, '0', String(qtyOf(testId))));
    /* restore: replace cart with the saved copy */
    cart.value = beforeCart;
    return checks;
  }));

  return results;
}

function behaviorOne(fn: string, label: string, run: () => TestCheck[]): TestResult {
  let checks: TestCheck[] = [];
  let crashed = false;
  try {
    checks = run();
  } catch (e) {
    crashed = true;
    checks.push({
      name: `${fn} רץ בלי לקרוס`,
      pass: false,
      detail: e instanceof Error ? e.message : String(e),
    });
  }
  if (!crashed) {
    checks.push({ name: `${fn} רץ בלי לקרוס`, pass: true });
  }
  return {
    id: `behavior:${fn}`,
    category: 'behavior' as const,
    label: `${fn} · ${label}`,
    area: 'התנהגות',
    checks,
  };
}

function verdict(name: string, pass: boolean, expected: unknown, got: unknown): TestCheck {
  return { name, pass, expected: String(expected), got: String(got) };
}

export function testButtons(): TestResult[] {
  return [...existenceCheck(), ...behaviorCheck()];
}

File: app_flutter/lib/test_harness/tests/buttons.dart
import 'package:buildsmart/state/dial_state.dart';
import 'package:buildsmart/test_harness/types.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// Button-action tests — toggle each FAB dial, switch personas, drill paths.
/// Pattern: save state → mutate → assert → restore.
List<TestResult> testButtons(WidgetRef ref) {
  return [
    _runOne(
      id: 'button:openDial-each',
      label: 'openDial — פתיחת כל אחד מ-5 ה-FABs',
      area: 'דיאל',
      run: () {
        final checks = <TestCheck>[];
        final before = ref.read(openDialProvider);
        for (final d in OpenDial.values) {
          ref.read(openDialProvider.notifier).state = d;
          final got = ref.read(openDialProvider);
          checks.add(TestCheck(
            name: 'דיאל "${d.name}" נפתח',
            pass: got == d,
            expected: d.name,
            got: got.name,
          ));
        }
        ref.read(openDialProvider.notifier).state = before;
        return checks;
      },
    ),
    _runOne(
      id: 'button:activePersona',
      label: 'activePersona — מעבר בין דמויות',
      area: 'BS',
      run: () {
        final checks = <TestCheck>[];
        final before = ref.read(activePersonaProvider);
        for (final p in ['manager', 'contractor', 'store', 'courier', 'worker']) {
          ref.read(activePersonaProvider.notifier).state = p;
          final got = ref.read(activePersonaProvider);
          checks.add(TestCheck(
            name: 'מעבר ל-$p',
            pass: got == p,
            expected: p,
            got: got ?? 'null',
          ));
        }
        ref.read(activePersonaProvider.notifier).state = null;
        checks.add(TestCheck(
          name: 'איפוס ל-null (חזרה לרשת 5 דמויות)',
          pass: ref.read(activePersonaProvider) == null,
          got: '${ref.read(activePersonaProvider)}',
        ));
        ref.read(activePersonaProvider.notifier).state = before;
        return checks;
      },
    ),
    _runOne(
      id: 'button:searchTool',
      label: 'searchTool — בחירת כלי חיפוש',
      area: 'חיפוש',
      run: () {
        final checks = <TestCheck>[];
        final before = ref.read(searchToolProvider);
        for (final t in SearchTool.values) {
          ref.read(searchToolProvider.notifier).state = t;
          checks.add(TestCheck(
            name: 'בחירת ${t.name}',
            pass: ref.read(searchToolProvider) == t,
            got: ref.read(searchToolProvider)?.name ?? 'null',
          ));
        }
        ref.read(searchToolProvider.notifier).state = null;
        checks.add(TestCheck(
          name: 'איפוס ל-null',
          pass: ref.read(searchToolProvider) == null,
        ));
        ref.read(searchToolProvider.notifier).state = before;
        return checks;
      },
    ),
    _runOne(
      id: 'button:resetAllDials',
      label: 'resetAllDials — מאפס את כל ה-state של ה-dial',
      area: 'דיאל',
      run: () {
        final checks = <TestCheck>[];
        // Snapshot
        final bd = ref.read(openDialProvider);
        final ap = ref.read(activePersonaProvider);
        final st = ref.read(searchToolProvider);
        // Make some state
        ref.read(openDialProvider.notifier).state = OpenDial.search;
        ref.read(activePersonaProvider.notifier).state = 'manager';
        ref.read(searchToolProvider.notifier).state = SearchTool.voice;
        // Reset
        resetAllDials(ref);
        checks.add(TestCheck(
          name: 'openDial → none',
          pass: ref.read(openDialProvider) == OpenDial.none,
          got: ref.read(openDialProvider).name,
        ));
        checks.add(TestCheck(
          name: 'activePersona → null',
          pass: ref.read(activePersonaProvider) == null,
          got: '${ref.read(activePersonaProvider)}',
        ));
        checks.add(TestCheck(
          name: 'searchTool → null',
          pass: ref.read(searchToolProvider) == null,
        ));
        // Restore
        ref.read(openDialProvider.notifier).state = bd;
        ref.read(activePersonaProvider.notifier).state = ap;
        ref.read(searchToolProvider.notifier).state = st;
        return checks;
      },
    ),
  ];
}

TestResult _runOne({
  required String id,
  required String label,
  required String area,
  required List<TestCheck> Function() run,
}) {
  var checks = <TestCheck>[];
  var crashed = false;
  try {
    checks = run();
  } on Object catch (e) {
    crashed = true;
    checks.add(TestCheck(
      name: 'הבדיקה רצה בלי לקרוס',
      pass: false,
      detail: '$e',
    ));
  }
  if (!crashed) {
    checks.add(const TestCheck(name: 'הבדיקה רצה בלי לקרוס', pass: true));
  }
  return TestResult(
    id: id,
    category: TestCategory.buttons,
    label: label,
    area: area,
    checks: checks,
  );
}
